WalletWasabi/Tor/Socks5/Pool/PoolItem.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace WalletWasabi.Tor.Socks5.Pool
+{
+	public enum PoolItemState
+	{
+		InUse,
+		FreeToUse,
+		ToDispose
+	}
+
+	/// <summary>
+	/// Pool item represents a single TCP connection to Tor SOCKS5 endpoint.
+	/// <para>Once a pool item is created, it is in <see cref="PoolItemState.InUse"/> state and the internal TCP connection is used immediately
+	/// to send an HTTP(s) request.</para>
+	/// <para>Then it is decided whether the pool item can be re-used for a next HTTP(s) request or not.</para>
+	/// </summary>
+	/// <remarks>Currently we re-use TCP connection to Tor SOCKS5 endpoint for HTTP requests but not for HTTPS requests.</remarks>
+	public class PoolItem
+	{
+		private static long Counter;
+
+		private bool _disposedValue;
+
+		/// <summary>
+		/// Creates a new instance of the object.
+		/// </summary>
+		/// <param name="client">TCP client connected to Tor SOCKS5 endpoint.</param>
+		/// <param name="allowRecycling">Whether it is allowed to re-use this Tor pool item.</param>
+		public PoolItem(TorConnection client, bool allowRecycling)
+		{
+			Id = Interlocked.Increment(ref Counter);
+			State = PoolItemState.InUse;
+			Client = client;
+			AllowRecycling = allowRecycling;
+		}
+
+		private object StateLock { get; } = new object();
+
+		public PoolItemState State { get; private set; }
+		private TorConnection Client { get; set; }
+		private bool AllowRecycling { get; }
+		private long Id { get; }
+
+		public TorConnection GetClient()
+		{
+			lock (StateLock)
+			{
+				Debug.Assert(State == PoolItemState.InUse);
```

Comment:
In order to fix the CodeFactor issue a call to Debug.Assert must provide a message in the second parameter describing the reason for the assert.

Reply:
I guess this guard can be removed now.